hoomd/mpcd/ReverseNonequilibriumShearFlow.h

```diff
@@ -0,0 +1,119 @@
+#define MPCD_REVERSE_NONEQUILIBRIUM_SHEAR_FLOW_H_
+
+#ifdef NVCC
+#error This header cannot be compiled by nvcc
+#endif
+
+#include "hoomd/ParticleGroup.h"
+#include "hoomd/Updater.h"
+#include <pybind11/pybind11.h>
+
+namespace hoomd
+    {
+namespace mpcd
+    {
+//! Reverse nonequilibrium shear flow updater
+/*!
+ * A flow is induced by swapping velocities in x direction based on particle position in
+ * y-direction.
+ */
+class PYBIND11_EXPORT ReverseNonequilibriumShearFlow : public Updater
+    {
+    public:
+    //! Constructor
+    ReverseNonequilibriumShearFlow(std::shared_ptr<SystemDefinition> sysdef,
+                                   std::shared_ptr<Trigger> trigger,
+                                   unsigned int num_swap,
+                                   Scalar slab_width,
+                                   Scalar p_target);
+
+    //! Destructor
+    virtual ~ReverseNonequilibriumShearFlow();
+
+    //! Apply velocity swaps
+    virtual void update(uint64_t timestep);
+
+    //! Set the maximum number of swapped pairs
+    void setNswap(unsigned int new_num_swap);
+
+    //! Set the slab width
+    void setSlabWidth(Scalar slab_width);
+
+    //! Set the target velocity
+    void setTargetMomentum(Scalar p_target);
+
+    //! Get max number of swaps
+    Scalar getNswap() const
+        {
+        return m_num_swap;
+        }
+
+    //! Get slab thickness
+    Scalar getSlabWidth() const
+        {
+        return m_slab_width;
+        }
+
+    //! Get target momentum
+    Scalar getTargetMomentum() const
+        {
+        return m_p_target;
+        }
+
+    //! Get summed exchanged momentum
+    Scalar getSummedExchangedMomentum() const
+        {
+        return m_momentum_exchange;
+        }
+
+    protected:
+    std::shared_ptr<mpcd::ParticleData> m_mpcd_pdata; //!< MPCD particle data
+    unsigned int m_num_swap;                          //!< maximum number of swaps
+    Scalar m_slab_width;                              //!< thickness of slabs
+    Scalar m_momentum_exchange;                       //!< summed momentum excange between slabs
+    Scalar2 m_lo_pos;                                 //!< position of bottom slab in box
+    Scalar2 m_hi_pos;                                 //!< position of top slab in box
+    unsigned int m_num_layer_lo;                      //!< number of particles in bottom slab
+    GPUArray<Scalar2> m_layer_lo;   //!< List of all particles (indices,momentum) in bottom slab
+                                    //!< sorted by momentum closest to +m_p_target
+    unsigned int m_num_layer_hi;    //!< number of particles in top slab
+    GPUArray<Scalar2> m_layer_hi;   //!< List of all particles (indices,momentum) in top slab sorted
+                                    //!< by momentum closest to -m_p_target
+    unsigned int m_num_staged;      //!< List of all staged particles (indices,momentum)
+    GPUArray<Scalar2> m_swap_queue; //!< stage momentum for swapping
```

Comment:
Let's also rename this one too. This consistency will make it easier to follow the code in future.
```suggestion
    GPUArray<Scalar2> m_particles_staged; //!< Particles staged for swapping
```